KAFKA-6106; Postpone normal processing of tasks within a thread until…

kamalcph opened the pull request:
… restoration of all tasks have completed.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

kamalcph committed:
KAFKA-6106; Postpone normal processing of tasks within a thread until restoration of all tasks have completed.

kamalcph commented:
@guozhangwang @mjsax 

Call for review. Should I have to clear the `resumedPartitions` when `onPartitionsRevoked` triggers ? (Not sure)

mjsax commented:
Thanks for the patch, @kamalcph 

I am not sure if this is sufficient to resolve the ticket. If I am not wrong, we only pause partitions for newly assigned tasks (cf `TaskManager#createTasks` L112-114). However, on restore we should pause all partitions during restore. Does this make sense?

mjsax commented:
\cc @bbejeck

kamalcph commented:
@mjsax Thanks for the pointer. Reassigned and new tasks with no state stores are not being paused. One option is to pause all the assigned partitions in `TaskManager#createTasks` until the stores are ready but some `StreamThreadTest` are failing. I'm working on it.

guozhangwang commented:
+1 to @mjsax 's comment.

@kamalcph I'd suggest you also consider do some code refactoring as we would only resume consumer on all partitions once all tasks have completed restoration. For example, `consumer.resume(resumed);` should be extracted outside of `updateNewAndRestoringTasks`.

smurakozi committed:
Fix in updateBrokerInfoInZk, exception is thrown if response was not OK.

smurakozi committed:
Fix in deleteLogDirEventNotifications - use correct path for deleted children

smurakozi committed:
Additional tests to improve test coverage of KafkaZkClient.

smurakozi committed:
Use move otherZkClient to KafkaZkClientTest

smurakozi committed:
Undo unintentional whitespace change in ZooKeeperTestHarness

smurakozi committed:
Fixes requested by reviewer

smurakozi committed:
Minor fixes requested by reviewer

mjsax reviewed (commented):
I guess this is still not sufficient. We need to make sure to pause all partitions. Atm, we only pause a partitions if its task has a state store.

cf. `TaskManager#createTasks` and `AssignedTasks#uninitializedPartitions`

I think we can remove `AssignedTasks#uninitializedPartitions` and pass to complete `assignment` that is passed into `TaskManager#createTasks` as parameter

mjsax commented:
Btw: please leave a comment after you update the PR. Otherwise, we might miss that you did update it.

kamalcph committed:
Pause all the partitions until the underlying state store is ready for all the active tasks.

kamalcph commented:
@mjsax, Sorry for the delay.  Last time, when I updated the code, I ran only the TaskManagerTest. After push, I realized that some of the StreamThreadTest cases are failing. So, I reverted the code immediately (this may caused the update in PR).

 I’ve updated this PR now. Still facing this exception in StreamThread test cases. 

    java.lang.IllegalStateException: No current assignment for partition topic1-1
	at org.apache.kafka.clients.consumer.internals.SubscriptionState.assignedState(SubscriptionState.java:259)
	at org.apache.kafka.clients.consumer.internals.SubscriptionState.pause(SubscriptionState.java:413)
	at org.apache.kafka.clients.consumer.MockConsumer.pause(MockConsumer.java:341)
	at org.apache.kafka.streams.processor.internals.TaskManager.createTasks(TaskManager.java:114)
	at org.apache.kafka.streams.processor.internals.StreamThreadTest.shouldReturnActiveTaskMetadataWhileRunningState(StreamThreadTest.java:717)

Previously, the partitions returned by `AssignedTasks#uninitializedPartitions` was empty. So, this error never occurred. I tried to subscribe (topic1) in mock consumer but it didn’t succeeded. Could you please help me out ?

mjsax commented:
I see. Thanks.

About the test. I am not sure, but maybe you need to call `thread.rebalanceListener.onPartitionsAssigned(assignedPartitions);` before `thread.taskManager().createTasks(assignedPartitions);`

kamalcph committed:
Topic assigned to MockConsumer before calling `onPartitionsAssigned` method.

kamalcph commented:
@mjsax Test cases are passed now. Please review it.

mjsax commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java:
```diff
@@ -323,18 +323,17 @@ void setConsumer(final Consumer<byte[], byte[]> consumer) {
      */
     boolean updateNewAndRestoringTasks() {
-        final Set<TopicPartition> resumed = active.initializeNewTasks();
+        active.initializeNewTasks();
```
I think we can cleanup `initializeNewTasks()` internally and change its return type to `void` ?

kamalcph replied:
Refactored it.

mjsax commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java:
```diff
@@ -323,18 +323,17 @@ void setConsumer(final Consumer<byte[], byte[]> consumer) {
         final Collection<TopicPartition> restored = changelogReader.restore(active);
 
-        resumed.addAll(active.updateRestored(restored));
+        active.updateRestored(restored);
```
Same for `updateRestored()`

kamalcph replied:
Refactored it.

mjsax reviewed (commented):
Some more comments. Did not look at the tests in detail yet.

Call for second review @guozhangwang @bbejeck @dguy @vvcephei

smurakozi committed:
Changes requeted by reviewer

smurakozi committed:
removed method updatedLeaderIsrAndControllerEpochs

junrao committed:
Revert "Undo unintentional whitespace change in ZooKeeperTestHarness"

This reverts commit 155ac5903aec0f682e19a0f1f17c1091f60c6da5.

junrao committed:
Revert "Fixes requested by reviewer"

This reverts commit ae51a15026e0dc10ef2df4bcafc85807dc8b4eb6.

junrao committed:
Revert "Fix in updateBrokerInfoInZk, exception is thrown if response was not OK."

This reverts commit 010310388725d6393a73e12c02dff4bb85cf2518.

junrao committed:
Revert "Use move otherZkClient to KafkaZkClientTest"

This reverts commit 54a32205b0fa7545d450104bc28aacf153b70cf6.

junrao committed:
Revert "removed method updatedLeaderIsrAndControllerEpochs"

This reverts commit 8100fd92afd6c4615bf7a6998a8a8eeb0d628d4e.

junrao committed:
Revert "Changes requeted by reviewer"

This reverts commit cacd377933ae0e7da0ed08dd33ab69fabde073c5.

junrao committed:
Revert "Minor fixes requested by reviewer"

This reverts commit 1ae4c1d6095a86266ba6d008711570df0eaf5682.

junrao committed:
Revert "Additional tests to improve test coverage of KafkaZkClient."

This reverts commit d0eb552a891ecd99150ddf4b89985525806c6e31.

junrao committed:
Revert "Fix in deleteLogDirEventNotifications - use correct path for deleted children"

This reverts commit af598af20828149bdd45336377e70d26360452fe.

smurakozi committed:
KAFKA-6111: Improve test coverage of KafkaZkClient, fix bugs found by new tests;

bbejeck commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java:
```diff
@@ -109,9 +109,9 @@ void createTasks(final Collection<TopicPartition> assignment) {
-        log.trace("Pausing partitions: {}", partitions);
-        consumer.pause(partitions);
+        // Pause all the partitions until the underlying state store is ready for all the active tasks.
+        log.trace("Pausing partitions: {}", assignment);
```
I'm wondering if this should be `debug` or not, same below when resuming assignment, although I don't feel too strongly about this.

kamalcph replied:
Earlier, It was logged as `trace` so left it as it is.

bbejeck commented on streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java:
```diff
@@ -626,13 +628,19 @@ public void shouldPunctuateActiveTasks() {
 
     @Test
-    public void shouldResumeConsumptionOfInitializedPartitions() {
+    public void shouldNotResumeConsumptionUntilAllStoresRestored() {
```
For this test would it be more clear to define a local strict mock for the consumer `Consumer<byte[], byte[]> consumer = EasyMock.createStrictMock(Consumer.class);` then get rid of lines 637-643?

kamalcph replied:
Updated the code to use strict mock.

bbejeck commented:
Left a couple of minor comments, assuming remaining comments from @mjsax are addressed, LGTM.

kamalcph commented:
@mjsax @bbejeck 

Thanks for reviewing! I have done rebase with trunk but somehow the commits got polluted in this PR. So, opened a new [PR](https://github.com/apache/kafka/pull/4651) and closing this one.
